### PR TITLE
Remove getInfoHostString and use connectable hostnames.

### PR DESCRIPTION
**Pull request by yuchenshi**
### Description

This is a refactor with a subtle behavior change to use correctly-formatted URLs with hostname fixes. This marks the start of a quest to fix IPv6 and DNS related problems.

When given a `0.0.0.0` hostname, we'll try to connect over `127.0.0.1` instead to make sure max client compatibility. `EmulatorRegistry.getInfo` now returns the transformed hostname by default, since this is usually the desirable behavior. A new helper method is added for creating an API client to an emulator -- a common enough pattern that worth standardizing. This PR also uses `EmulatorRegistry.url` for most URL constructions and has removed `getInfoHostString`, which causes more confusions than it helps.

### Scenarios Tested

Unit test added for the new `connectableHostname` helper.

Basic manual testing like starting the emulators. 

I'll rely on our CI to catch any remaining errors.

### Sample Commands

N/A

### CODE REVIEW

File: src/commands/emulators-start.ts [hunk 1]
```diff
@@ -95,22 +95,23 @@ function printEmulatorOverview(options: any): void {
...
+          url.pathname = `/${emulator}`;
+          uiLink = stylizeLink(url.toString());
+        }
 
-        return [
-          emulatorName,
-          EmulatorRegistry.getInfoHostString(info),
-          isSupportedByUi && uiInfo ? stylizeLink(`${uiUrl}/${emulator}`) : clc.blackBright("n/a"),
-        ];
+        return [emulatorName, info.host, info.port, uiLink];
```
Comment (bkendall) [inline src/commands/emulators-start.ts:109]: I'm certain the reason we had the full links before was so that they were clickable from a terminal - I remember it being a specific request! Could we maintain that?

File: src/emulator/commandUtils.ts [hunk 2]
```diff
@@ -276,17 +276,19 @@ function processKillSignal(
         const pids: number[] = [];
 
         const emulatorsTable = new Table({
-          head: ["Emulator", "Host:Port", "PID"],
+          head: ["Emulator", "Host", "Port", "PID"],
```
Comment (bkendall) [inline src/emulator/commandUtils.ts:279]: Similar here - this will make things un-clickable. Could we maintain that?

File: src/emulator/controller.ts [hunk 3]
```diff
@@ -502,14 +501,7 @@ export async function startAll(
...
     experiments.assertEnabled("webframeworks", "emulate a web framework");
-    const emulators: EmulatorInfo[] = [];
-    if (experiments.isEnabled("webframeworks")) {
-      for (const e of EMULATORS_SUPPORTED_BY_UI) {
-        const info = EmulatorRegistry.getInfo(e);
-        if (info) emulators.push(info);
-      }
-    }
-    await prepareFrameworks(targets, options, options, emulators);
+    await prepareFrameworks(targets, options, options);
```
Comment (bkendall) [inline src/emulator/controller.ts:504]: We will probably need to explicitly test that the frameworks stuff works. It looks fine, but it's sensitive to disruption right now.

File: src/emulator/pubsubEmulator.ts [hunk 4]
```diff
@@ -25,7 +25,7 @@ interface Trigger {
 }
 
 export class PubsubEmulator implements EmulatorInstance {
-  pubsub: PubSub;
+  private _pubsub: PubSub | undefined;
```
Comment (bkendall) [inline src/emulator/pubsubEmulator.ts:28]: it's redundant to prefix private fields with `_`. Please omit.

**Review by bkendall — CHANGES_REQUESTED** [review]
I think we'll want to manually test some webframeworks stuff here for sanity.

File: src/emulator/pubsubEmulator.ts (hunk 4 above)
Comment (yuchenshi) [reply src/emulator/pubsubEmulator.ts:28]: The getter is also called `pubsub` and we'd have a naming conflict if `_` were to be removed.

File: src/commands/emulators-start.ts (hunk 1 above)
Comment (yuchenshi) [reply src/commands/emulators-start.ts]: Good catch! I've restored the behavior. Added some comments too for background.

File: src/emulator/commandUtils.ts (hunk 2 above)
Comment (yuchenshi) [reply src/emulator/commandUtils.ts]: Good catch! I've restored the behavior. Added some comments too for background.

File: src/emulator/controller.ts (hunk 3 above)
Comment (yuchenshi) [reply src/emulator/controller.ts]: As discussed offline, I've reverted my changes to frameworks and added a comment on the code in question.

File: src/emulator/controller.ts [hunk 5]
```diff
@@ -505,6 +505,8 @@ export async function startAll(
     const emulators: EmulatorInfo[] = [];
     if (experiments.isEnabled("webframeworks")) {
       for (const e of EMULATORS_SUPPORTED_BY_UI) {
+        // TODO: Double check if this actually works -- we're early in the startup
```
Comment (christhompsongoogle) [inline src/emulator/controller.ts:508]: nit: add name/ldap

File: src/emulator/functionsEmulator.ts [hunk 6]
```diff
@@ -1149,41 +1132,11 @@ export class FunctionsEmulator implements EmulatorInstance {
...
-    const pubsubEmulator = this.getEmulatorInfo(Emulators.PUBSUB);
-    if (pubsubEmulator) {
-      const pubsubHost = formatHost(pubsubEmulator);
-      process.env.PUBSUB_EMULATOR_HOST = pubsubHost;
-    }
 
+    setEnvVarsForEmulators(envs);
     const eventarcEmulator = this.getEmulatorInfo(Emulators.EVENTARC);
     if (eventarcEmulator) {
+      // TODO: Why do we set the protocol here, but no protocol during emulators:exec?
```
Comment (christhompsongoogle) [inline src/emulator/functionsEmulator.ts:1139]: nit: add name/ldap

**Comment by christhompsongoogle** [comment]
Question: What's the semantic difference between "isRunning" and "getInfo" - is info populated before isRunning? Why switch from one to the other?

**Review by christhompsongoogle — APPROVED** [review]

File: src/emulator/controller.ts (hunk 5 above)
Comment (yuchenshi) [reply src/emulator/controller.ts:508]: I don't know the right contact yet but I've started a thread for it internally. If I cannot identify a contact, then I'll just create an internal bug for this TODO.

**Comment by yuchenshi** [comment]
> Question: What's the semantic difference between "isRunning" and "getInfo" when checking if an Emulator is running/active/enabled - is info populated before isRunning? Why switch from one to the other?

It is more or less equivalent, and there's no actual difference in timing. But `isRunning` is preferred because it captures the intention better.

Deeper reason: When using `getInfo`, if later one needs to construct a URL, they will naturally get the `host` and `port` from the info and try simple string interpolation, which works most of the time but fails in the edge cases in PR description. It is much safer to just do a boolean test and use `EmulatorRegistry.url` or `EmulatorRegistry.client` later if needed.

Future consideration: I'm about to create follow-up PRs to change the way emulator listens on hosts (with DNS in mind) and it is necessary to audit the situations where the `host` is absolutely needed. Less `getInfo` means less places to track down and potentially modify later.

Side benefit: `isRunning` does less work than `getInfo` -- a bit more performance, less moving pieces.

**Review by bkendall — APPROVED** [review]
